MINOR: Add LEAVE_GROUP_EPOCH to GroupMetadataManager

kirktrue opened the pull request:
Replacing the use a hardcoded -1 with a constant (`LEAVE_GROUP_EPOCH`) that provides more clarity. Since static members also have a magic number (-2), this will motivate future commits to use constants instead of hardcoded values.

kirktrue committed:
[MINOR] Add LEAVE_GROUP_EPOCH to GroupMetadataManager

Replacing the use a hardcoded -1 with a constant (LEAVE_GROUP_EPOCH)
that provides more clarity. Since static members also have a magic
number (-2), this will motivate future commits to use constants instead
of hardcoded values.

vamossagar12 reviewed (approved):
Thanks for this PR @kirktrue ! I can see easily see how this can be extended for static membership case. I had just one question(LGTM o/w) =>

1. Do [these](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1368-L1370) lines also qualify for having -1 replaced with `LEAVE_GROUP_EPOCH`? In this case, it appears to be a member leave group situation.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -131,6 +131,11 @@
+    /**
+     * A member epoch of <code>-1</code> means that the member wants to leave the group.
+     */
+    public static final int LEAVE_GROUP_EPOCH = -1;
```
I wonder if we should rather define this on in `ConsumerGroupHeartbeatRequest`. What do you think?

kirktrue replied:
Makes sense.

kirktrue replied:
Done.

kirktrue committed:
Moved/renamed 'leave group epoch' based on feedback

GroupMetadataManager.LEAVE_GROUP_EPOCH

is now:

ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH

kirktrue committed:
Update GroupMetadataManager.java

Setting 'previous member epoch' and 'target member epoch' from hard-coded -1 to LEAVE_GROUP_MEMBER_EPOCH.

kirktrue commented:
> Thanks for this PR @kirktrue ! I can see easily see how this can be extended for static membership case. I had just one question(LGTM o/w) =>
> 
> 1. Do [these](https://github.com/apache/kafka/blob/trunk/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1368-L1370) lines also qualify for having -1 replaced with `LEAVE_GROUP_EPOCH`? In this case, it appears to be a member leave group situation.

I've updated the code to use the constant in place of `-1`. Thanks!

kirktrue commented:
@dajac @vamossagar12 Please take another look. Thanks!

vamossagar12 commented:
Thanks @kirktrue , LGTM (I had already approved).

dajac reviewed (approved):
LGTM, thanks
